### PR TITLE
ML-036/add view list

### PR DESCRIPTION
**Pull request by twill11**
# Description

Add aasm
Add view list entity
Add states to view list
Add view list specs
Improve ratings 
Add indexes on validations

## Trello card link

https://trello.com/c/nfUXSO8F/10-ml-036-add-view-items

# Checklist:

- [x] My code fixes the problem or adds required functionality
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings
- [x] I have added tests that prove that my feature works
- [x] New and existing unit tests pass locally with my changes

# Additional requirements:

- [x] Does your code require any migrations to be ran?
- [x] Does your code require installing any new gems or system dependencies?
- [x] Does your code require any changes to current deployment process?

# Additional Information

### CODE REVIEW

File: app/controllers/api/v1/ratings_controller.rb [hunk 1]
```diff
@@ -1,18 +1,15 @@
 module Api
   module V1
     class RatingsController < ApplicationController
-      before_action :set_rating, only: %i[show update destroy]
+      before_action :authenticate_user!
```
Comment (WhatWasThatt) [inline app/controllers/api/v1/ratings_controller.rb:4]: I think you can put it to the ApplicationController

File: app/service/view_list_service.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
+class ViewListService
+  def initialize(class_name, params, current_user)
+    @class_name = class_name
+    @params = params
+    @current_user = current_user
+  end
+
+  def call
+    class_name.where(user_id: current_user.id).paginate(page: params[:page]).order(created_at: :desc)
```
Comment (WhatWasThatt) [inline app/service/view_list_service.rb:9]: I think you could sort by watch status

File: spec/factories/view_lists.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+# Foreign Keys
+#
+#  fk_rails_...  (movie_id => movies.id)
+#  fk_rails_...  (user_id => users.id)
+#
+FactoryBot.define do
+  factory :view_list do
+    association :user
+    association :movie
+  end
```
Comment (WhatWasThatt) [inline spec/factories/view_lists.rb:27]: you can use traits for watching status

File: spec/models/view_list_spec.rb [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+#  fk_rails_...  (user_id => users.id)
+#
+require 'rails_helper'
+
+RSpec.describe ViewList, type: :model do
+  subject(:view_list) { create :view_list }
+
+  context 'with valid attributes' do
+    it { expect(view_list).to be_valid }
+  end
```
Comment (WhatWasThatt) [inline spec/models/view_list_spec.rb:30]: and test statuses

File: Gemfile [hunk 5]
```diff
@@ -23,6 +23,10 @@ gem 'rswag-ui'
 # gem 'image_processing', '~> 1.2'
 
 # Reduces boot times through caching; required in config/boot.rb
+gem 'aasm', '~> 5.2'
+
+gem 'after_commit_everywhere', '~> 1.0'
```
Comment (WhatWasThatt) [inline Gemfile:28]: tell me please, why do you need this gem now?

**Comment by WhatWasThatt** [comment]
and please add request tests

File: app/controllers/api/v1/ratings_controller.rb (hunk 1 above)
Comment (twill11) [reply app/controllers/api/v1/ratings_controller.rb:4]: are you sure about put authenticate_user in application controller? i think guest user can see movies and persons without registration

Comment (WhatWasThatt) [reply app/controllers/api/v1/ratings_controller.rb:4]: we can add skip_before_action for movie and person controllers

**Review by WhatWasThatt — APPROVED** [review]
good job

File: config/routes.rb [hunk 6]
```diff
@@ -15,6 +15,7 @@
         resource :permissions
       end
       resources :tags
+      resources :view_lists
```
Comment (dmitrymalllpr) [inline config/routes.rb:18]: users/{id}/view_lists

**Review by WhatWasThatt — DISMISSED** [review]
good job

File: app/service/rating_service.rb [hunk 7]
```diff
@@ -1,17 +1,17 @@
...
+    class_name.where(user: user).paginate(page: params[:page]).order("rating #{order}")
   end
 
   private
 
-  attr_accessor :class_name, :params, :current_user
+  attr_reader :class_name, :params, :user
 
   def order
     params[:order].presence || 'asc'
```
Comment (dmitrymalllpr) [inline app/service/rating_service.rb:17]: Rework please. Don t pass params

**Review by dmitrymalllpr — CHANGES_REQUESTED** [review]

**Review by WhatWasThatt — DISMISSED** [review]

**Review by WhatWasThatt — APPROVED** [review]

**Pull request merged**